boards: add the Adafruit Trinket M0.

nzmichaelh opened the pull request:
The Trinket M0 is a tiny SAMD21 based board with USB, UART, SPI, and
an on-board RGB LED.

Signed-off-by: Michael Hope <mlhx@google.com>

ydamigos commented on boards/arm/adafruit_trinket_m0/adafruit_trinket_m0.dts:
```diff
@@ -0,0 +1,77 @@
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	apa102@0 {
```
DT binding for this device is missing.

ydamigos replied:
Ok, PR #6209 provides the binding

nzmichaelh replied:
Yip :)

nzmichaelh commented:
I asked Adafruit and they said it was OK to use their photo for the board image.  Updated.

galak commented on boards/arm/adafruit_trinket_m0/dts.fixup:
```diff
@@ -0,0 +1,54 @@
+#define CONFIG_GPIO_SAM0_PORTA_LABEL ATMEL_SAM0_GPIO_41004400_LABEL
```
Lets refactor the atmel sam0 dts.fixup into arch/arm/soc/

nzmichaelh replied:
Done.  Note that the samd/ version wasn't getting picked up so I moved it to samd21/

galak reviewed (changes requested):
Looks good overall, lets refactor dts.fixup to remove duplication (see commit d76858f6ae91c325c59ff206977e60be2f384251 for an example)

dbkinder commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,147 @@
+- 256 KiB flash memory and 32 KiB of RAM
+- Internal trimmed 8 MHz oscillator
+- A user LED
+- A RGB DotStar LED
```
``An RGB``   (and yes, English rules for using ``a`` vs. ``an`` are tricky, based on the sound of the next word)

nzmichaelh replied:
Done

dbkinder commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,147 @@
+
+Other hardware features are not currently supported by Zephyr.
+
+The default configuration can be found in the Kconfig
```
``in the Kconfig file``

nzmichaelh replied:
Done

dbkinder commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,147 @@
+Connections and IOs
+===================
+
+The `Adafruit Learn`_ site has detailed information about the board
```
More descriptive if you write this as (and adjust the label name at the end of the doc): 

    The `Adafruit Trinket M0 Learn site`_ has ...

nzmichaelh replied:
Done

dbkinder commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,147 @@
+*************************
+
+The Trinket M0 ships the BOSSA compatible UF2 bootloader.  The
+bootloader can be entered by quickly pressing the reset button twice.
```
You use the term ``tapping`` below, I like that better since it implies a press and release:

    by quickly tapping the reset button twice.

nzmichaelh replied:
Done

dbkinder commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,147 @@
+
+#. Connect the Trinket M0 to your host computer using USB
+
+#. Connect a 3.3 V USB to serial adapter to the board and to the host
```
maybe add:

    (See the `Serial Port`_ section above for the board's pin connections.)

nzmichaelh replied:
Done

dbkinder commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,147 @@
+
+.. target-notes::
+
+.. _Adafruit Learn:
```
update this label with the suggestion above (where it's referenced)

nzmichaelh replied:
Done

dbkinder commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,147 @@
+Overview
+********
+
+The Adafruit Trinket M0 is a tiny (27 mm x 15 mm) development board
```
Maybe add ``ARM`` before ``development board`` (to let folks know the architecture early on)

nzmichaelh replied:
Done

dbkinder reviewed (changes requested):
thanks for the doc.  some suggested edits

dbkinder reviewed (approved):
Thanks!

nzmichaelh commented:
@galak I've refactored the dts fixup to arch/arm/soc/...  I also dropped the LED strip driver as the new SPI master checks will fail until #5865 is merged

henrikbrixandersen commented:
Shouldn't `dts.fixup` also include definitions for SPI (`CONFIG_SPI_SAM0_SERCOM0_BASE_ADDRESS` and friends)?

nzmichaelh commented:
Hi @henrikbrixandersen my SPI in Device Tree (#5865) PR needs to land first.

I'm hoping that this PR gets approved then we can continue with the SPI related changes in the SPI PR.

henrikbrixandersen commented:
Hi,

Ok, thank you. I have a port for the Adafruit Feather M0 Basic Proto ready - but waiting for this PR to land before submitting it since it too would need a refactoring af the dts.fixup.

nzmichaelh commented:
Cool!  If you happen to get a Feather M0 Adalogger in the future then you might be interested in the #5528 SD card driver.

henrikbrixandersen commented on boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst:
```diff
@@ -0,0 +1,149 @@
+.. _adafruit_trinket_m0:
+
+Adafruit Tinket M0
```
s/Tinket/Trinket/

nzmichaelh replied:
Good spotting, fixed and ran a spell check on the file.

galak commented on samples/drivers/led_ws2812/src/main.c:
```diff
@@ -19,7 +19,7 @@
  * Number of RGB LEDs in the LED strip, adjust as needed.
  */
-#define STRIP_NUM_LEDS 12
+#define STRIP_NUM_LEDS 1
```
Is this change related to the board code?  At min we need a comment in commit message about it.  Ideally it would be a separate commit.

nzmichaelh replied:
Oops, fixed.

galak commented on arch/arm/soc/atmel_sam0/samd21/dts.fixup:
```diff
@@ -1,3 +1,8 @@
+#define FLASH_DEV_BASE_ADDRESS		ATMEL_SAM0_NVMCTRL_41004000_BASE_ADDRESS_0
+#define FLASH_DEV_NAME			ATMEL_SAM0_NVMCTRL_41004000_LABEL
+
 #define CONFIG_GPIO_SAM0_PORTA_LABEL ATMEL_SAM0_GPIO_41004400_LABEL
```
Can you pull the dts.fixup move/additions into their own commit in the PR.  I'd like them separated from the board code commit if possible.

nzmichaelh replied:
Done as #6305

galak reviewed (changes requested):
Some minor cleanups / refactoring and question about change to led_ws2812 driver.

nzmichaelh committed:
boards: add the Adafruit Trinket M0.

The Trinket M0 is a tiny SAMD21 based board with USB, UART, SPI, and
an on-board RGB LED.

Signed-off-by: Michael Hope <mlhx@google.com>

galak commented:
Hmm, from shippable:

../../../../../../../include/arch/arm/cortex_m/cmsis.h:132:2: error: #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
 #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
  ^~~~~

galak commented:
> Hmm, from shippable:
> 
> ../../../../../../../include/arch/arm/cortex_m/cmsis.h:132:2: error: #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
> #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
> ^~~~~

Looks like we just need to update against master to get the dts.fixup

nzmichaelh commented:
@galak I've rebased against master.
